### DISCUSSION TITLE
[react] `ComponentProps` should return `{}` for prop-less components?

### DISCUSSION THREAD

**Discussion by Josh-Cena** [DefinitelyTyped/DefinitelyTyped#60766] [Issues with a @types package]
In Docusaurus, we have a process called [wrap swizzling](https://docusaurus.io/docs/swizzling#wrapping), which would produce a wrapper of a component. The component will receive all props that used to be received by the original component. To facilitate this process, we have a CLI that automatically produces the following code:

```ts
import React, {ComponentProps} from 'react';
import AnnouncementBar from '@theme-original/AnnouncementBar';
import type AnnouncementBarType from '@theme/AnnouncementBar';

type Props = ComponentProps<typeof AnnouncementBarType>;

export default function AnnouncementBarWrapper(props: Props): JSX.Element {
  return (
    <>
      <AnnouncementBar {...props} />
    </>
  );
}
```

However, if `AnnouncementBarType` is a function like `() => JSX.Element`, `Props` will have type `unknown`. This is because `(() => JSX.Element) extends ((p: infer P) => JSX.Element) ? P : never` is `unknown`. Spreading `unknown` as JSX attribute used to work, but with TS 4.7, this no longer works.

[Playground](https://www.typescriptlang.org/play?ts=4.7.2#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgIilQ3wG4AocgEyTQBsUi5MBXAOw2AjbgGVckAYVxgAFAEoAXHABSvABoA6AKJ0kIJGxgUYATzBI4ABRxgAznAC8iYjEXDw3TTBMRzAHj0GImPgIdgAHwU5KwcMFw8AOpQKGAGUAGiYKZm0q7m4nAA3uRwcEQwLFA87vwaATmK1SluZgC+cAD0weT15EA). Note that this produces another error "Type 'unknown' is not assignable to type 'IntrinsicAttributes'." which is caused by the same issue and can be fixed once `ComponentProps` returns `{}`. It has gone away unnoticed in the past because we had a convenient ambient declaration for `@theme-original/*`, causing `AnnouncementBar` to have type `any`.

Is it possible for `ComponentProps` to special-case for prop-less components, and produce `{}` instead? This is what happens at runtime anyway. The impact of this goes beyond our use-case: it means any type produced by `ComponentProps` can be safely spread into another component, which is desirable.

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment]
To be clear: `{}` means any non-nullish value in TypeScript. So `string` or `number` would be assignable to `{}`. Is this what you're asking for?

**Reply by Josh-Cena** [reply]
I'm aware of that. But TS—whether intentionally or unintentionally—allows spreading `{}`. I just want something that can be spread as JSX attributes. I'll be equally happy if it produces `Record<never, never>` instead.

**Reply by eps1lon** [reply]
> I just want something that can be spread as JSX attributes

That's fine. But we have to consider implications for other library users as well. And I don't think a prop-less component should have a props shape that allows just any props.

For example, `function Component() {}` doesn't allow any props but you're suggesting it should allow any?

**Reply by Josh-Cena** [reply]
That's a good point.

1. Under the current situation, `function Component() {}` already has `ComponentProps` as `unknown`, which to me... looks like it's saying "I don't care what's being passed in because I don't read it anyway".
2. The element `<Component />`, at runtime, would pass `{}` (empty object) as props to `Component`, which means in practice the props a component receives would never not be an object.
3. [`{}`](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgIilQ3wG4AoczAVwDsNgJa4AxCCACjAC44BvAL4BKfuThwiMalGYAeACbAAbnAD0APgoDKsthDgoAvHwCMAtevJA) and [`Record<never, never>`](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgIilQ3wG4AoczAVwDsNgJa4AxCCACjAC5Ek1oAEwA8tJADckUADRwxkqAD4AlHADe5OHCIxqUZsMHBxcAPSKKAX0rC2EOCgC8agIyWzi8kA) would both prevent excess properties, because in JSX, it would be intersected with `IntrinsicAttributes`.
4. I can't think of a case where type safety could be _circumvented_, since both of them are subtypes of `unknown` anyway, which means if you, in the current world, have something like `props: ComponentProps<typeof Component>` and you somehow pass `props` around (maybe spreading it into the component like we previously were able to), `props` can only accept a stricter subset of values after this change.
